bridge/tracker/tracker.go

```diff
@@ -0,0 +1,390 @@
+		return
+	}
+
+	t.logger.Info("querying events within block range", "from", fromBlock, "to", toBlock)
+
+	//	fetch logs
+	logs := t.queryEvents(fromBlock, toBlock)
+	t.logger.Info("matched events", "num", len(logs))
+
+	//	notify each matched log event
+	t.notify(logs...)
+}
+
+//	calculateRange determines the next range of blocks
+//	to query for events.
+func (t *Tracker) calculateRange(header *ethHeader) (from, to uint64, ok bool) {
+	//	extract block number from header field
+	//	TODO: polygon-edge header
+	latestHeight := big.NewInt(0).SetUint64(header.Number)
+
+	//	or: TODO ethereum header
+	//latestHeight, _ := types.ParseUint256orHex(&header.Number)
+
+	//	check if block number is at required depth
+	confirmations := big.NewInt(0).SetUint64(t.confirmations)
+	if latestHeight.Cmp(confirmations) < 0 {
+		//	block height less than required
+		t.logger.Debug(
+			"not enough confirmations",
+			"current", latestHeight.Uint64(),
+			"required", confirmations.Uint64())
+
+		return
+	}
+
+	/*	right bound */
+	toBlock := big.NewInt(0).Sub(latestHeight, confirmations)
+
+	/*	left bound */
+	fromBlock := t.loadLastBlock()
+	if fromBlock == nil {
+		fromBlock = toBlock
+	} else {
+		//	increment - last block number was already queried
+		fromBlock = fromBlock.Add(fromBlock, big.NewInt(1))
+	}
+
+	if toBlock.Cmp(fromBlock) < 0 {
+		//	no need to perform any query
+		//	as left bound is higher than the right
+		return
+	}
+
+	return fromBlock.Uint64(), toBlock.Uint64(), true
+}
+
+//	loadLastBlock returns the block number of the last
+// 	block processed by the tracker (if available).
+func (t *Tracker) loadLastBlock() *big.Int {
+	//	read from db
+	bytesLastBlock, err := t.db.Get([]byte(lastQueriedBlockNumber), nil)
+	if err != nil {
+		t.logger.Error("cannot read db", "err", err)
+
+		return nil
+	}
+
+	//	parse
+	lastBlock, ok := big.NewInt(0).SetString(string(bytesLastBlock), 10)
+	if !ok {
+		return nil
+	}
+
+	return lastBlock
+}
+
+//	saveLastBlock stores the number of the last block processed
+//	into the tracker's database.	.
+func (t *Tracker) saveLastBlock(blockNumber *big.Int) error {
+	//	store last processed block number
+	if err := t.db.Put(
+		[]byte(lastQueriedBlockNumber),
+		[]byte(blockNumber.String()),
+		nil,
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//	queryEvents collects all events on the rootchain that occurred
+//	between blocks fromBlock and toBlock (inclusive).
+func (t *Tracker) queryEvents(fromBlock, toBlock uint64) []*web3.Log {
+	queryFilter := &web3.LogFilter{}
+
+	//	set range of blocks to query
+	queryFilter.SetFromUint64(fromBlock)
+	queryFilter.SetToUint64(toBlock)
+
+	//	set smart contract addresses
+	queryFilter.Address = []web3.Address{
+		web3.HexToAddress(PoCSC),
+		web3.HexToAddress(AnotherEventSC),
+		web3.HexToAddress(ThirdEventSC),
+	}
+
+	//	set relevant topics (defined in contracts)
+	queryFilter.Topics = [][]*web3.Hash{
+		{
+			&topicPoCEvent,
+			&topicAnotherEvent,
+		},
+	}
```

Comment:
This should be a separate method, like `setupQueryFilter` or something

Reply:
[Exported](https://github.com/0xPolygon/polygon-edge/pull/386/commits/07ac0e157de64918862ae67a75644fb2aa69a1f6)